app/policies/subject_rule_effect_policy.rb

```diff
@@ -7,57 +7,85 @@ def resolve
     end
   end
 
+  # record passed in from controller is a workflow
   def index?
-    # record is a workflow from the controller
     if credential.admin?
       true
     else
-      credential.project_ids.include?(record.project_id)
+      valid_workflow?(record, credential)
     end
   end
 
+  # record passed in from controller is a workflow
+  def show?
+    index?
+  end
+
+  # record passed in from controller is a workflow
+  def new?
+    index?
+  end
+
+  # record passed in from controller is a subject_rule_effect
   def create?
-    # record is a workflow from the controller
-    if credential.admin?
-      true
-    else
-      # temporarily disable the ability for users to create/update
-      # subject rule effects
-      false
-    end
+    update?
   end
 
+  # record passed in from controller is a subject_rule_effect
   def edit?
-    # record is a subject_rule_effect from the controller
     if credential.admin?
       true
     else
-      subject_rule_project_id = record.subject_rule.workflow.project_id
-      credential.project_ids.include?(subject_rule_project_id)
+      valid_associated_project?(record)
     end
   end
 
+  # record passed in from controller is a subject_rule_effect
   def update?
-    # record is a subject_rule_effect from the controller
-    if credential.admin?
-      true
-    else
-      # this is a good place to check the defined subect set id in the
-      # rule effect belongs to the scoped credential project ids using the API client
+    return false unless valid_credentials?
+    return true if credential.admin?
+
+    return false unless valid_associated_project?(record)
 
-      # temporarily disable the ability for users to create/update
-      # subject rule effects
-      false
+    if record.effect.is_a?(Effects::AddSubjectToSet)
+      valid_subject_set?(record)
+    elsif record.effect.is_a?(Effects::AddSubjectToCollection)
+      valid_collection?(record)
+    else
+      true
```

Comment:
using this variable from the above change allows me to understand how your earlier check plays out here with the other set / collection checks.
```suggestion
      user_has_project_access # should only be true if we make it here
```

You could also replace the variables with an explanatory comment, i didn't understand why the true value was being used but now i see it works with the earlier `valid_associated_project?` check (or i'm really missing the point).

Reply:
I've renamed `valid_associated_project?` to `user_has_project_access?` , and added a comment for clarity. I think that should put us in good shape :)